src/MainWindow.vala

```diff
@@ -30,7 +31,22 @@ namespace Terminal {
         private Granite.AccelLabel open_in_browser_menuitem_label;
 
         private HashTable<string, TerminalWidget> restorable_terminals;
-        private bool is_fullscreen = false;
+        private bool is_fullscreen {
+            get {
+                return header.decoration_layout_set;
+            }
+
+            set {
+                if (value) {
+                    fullscreen ();
+                    header.decoration_layout_set = true;
+                    //TODO Show Toast like Chrome? or additional unfullscreen button?
+                } else {
+                    unfullscreen ();
+                    header.decoration_layout_set = false;
+                }
+            }
+        }
```

Comment:
you can make this query the window state, that matches the behaviour of Gtk.Window.fullscreened in Gtk4.

```suggestion
        private bool is_fullscreen {
            get {
                if (!get_realized ()) {
                    return false;
                }
                
                return FULLSCREEN in get_window ().get_state ();
            }

            set {
                if (value) {
                    fullscreen ();
                } else {
                    unfullscreen ();
                }
            }
        }
```

then you can do `bind_property ("is-fullscreen", header, "decoration-layout-set", INVERT_BOOLEAN)`.

Reply:
Yes, that's better - thanks.  I was aware that fullscreen () cannot be guaranteed to actually fullscreen the window.  That is a nice solution if it works.  I am not sure that `bind_property` works with a property that is not backed by a private variable though - `notify` doesn't.

Reply:
GLib.Binding uses notify underneath, so yeah, i believe you will need to call `notify_property()` in a window_state_event handler for that.

Reply:
I tried getting this to work but there seems to be some weird timing issues where changing `decoration_layout_set` in `configure-event` has no effect.  As this is all going to have to be rewritten for Gtk4 anyway I do not want to spend more time on it.  Is it OK to go with the current solution?

Reply:
Fine for me.